### PR TITLE
Erase i2c peripheral instances

### PR DESCRIPTION
**Pull request by bugadani**
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [x] I have added necessary changes to user code to the [Migration Guide](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/MIGRATING-0.21.md).
- [x] My changes are in accordance to the [esp-rs API guidelines](https://github.com/esp-rs/esp-hal/blob/main/documentation/API-GUIDELINES.md)

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/documentation/CONTRIBUTING.md) and followed its instructions.

### CODE REVIEW

File: esp-hal/src/i2c.rs [hunk 1]
```diff
@@ -518,65 +520,89 @@ where
...
+        frequency: HertzU32,
+    ) -> Self {
+        Self::new_with_timeout(i2c, sda, scl, frequency, None)
+    }
 
-        self.peripheral.setup(self.frequency, self.timeout);
+    /// Create a new I2C instance with a custom timeout value.
+    /// This will enable the peripheral but the peripheral won't get
+    /// automatically disabled when this gets dropped.
+    pub fn new_with_timeout<
```
Comment (bugadani) [inline esp-hal/src/i2c.rs:569]: This constructor is terrible and in the current implementation it's pretty easy to introduce a `.with_timeout()` instead. Anybody knows why this wasn't done initially, e.g. was the timeout setting more intertwined in the past with `configure_clock`'s contents?

Comment (bugadani) [reply esp-hal/src/i2c.rs]: Yeah I guess the half_cycle calculation makes sense and it's needed for the timeout. Although I don't quite understand what happens here - why do we set timeout_enable = true even if timeout is None?

File: esp-hal/src/i2c.rs [hunk 2]
```diff
@@ -496,20 +497,20 @@ where
...
             phantom: PhantomData,
             frequency,
-            timeout,
+            timeout: None,
         };
 
+        PeripheralClockControl::reset(i2c.i2c.peripheral());
+        PeripheralClockControl::enable(i2c.i2c.peripheral());
+
+        // TODO: implement with_pins et. al.
```
Comment (Dominaezzz) [inline esp-hal/src/i2c.rs:513]: Is there a tracking issue for this `with_pins` idea?

File: esp-hal/src/i2c.rs [hunk 3]
```diff
@@ -2240,44 +2186,37 @@ where
...
-            Command::Read { ack_value, length } => {
-                cmd.write(|w| {
-                    w.opcode().read();
-                    w.ack_value().bit(ack_value == Ack::Nack);
-                    w.byte_num().bits(length);
-                    w
-                });
-            }
+
+    match command {
```
Comment (Dominaezzz) [inline esp-hal/src/i2c.rs:2190]: I'm a little surprised you didn't move the `match` into the `cmd.write(|w| .....)` lambda. :stuck_out_tongue:

**Review by Dominaezzz — APPROVED** [review]
LGTM

**Review by bjoernQ — APPROVED** [review]
LGTM

File: esp-hal/src/i2c.rs (hunk 3 above)
Comment (bugadani) [reply esp-hal/src/i2c.rs:2190]: oops

**Pull request merged**